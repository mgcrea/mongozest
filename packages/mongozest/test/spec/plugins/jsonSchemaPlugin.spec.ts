import createMongo, {Model, jsonSchemaPlugin} from './../../../src';
import {basename} from 'path';
import {kebabCase} from 'lodash';
import {Db} from 'mongodb';

const DB_NAME = getDbName(__filename);

const mongo = createMongo();

class Test extends Model {
  static schema = {
    name: {bsonType: 'string', required: true, someInvalidProp: true},
    avatar: {bsonType: 'object', properties: {fileName: 'string', size: 'long'}},
    list: {
      bsonType: 'array',
      items: {
        oneOf: [{bsonType: 'string'}, {bsonType: 'int'}]
      }
    },
    nestedObject: {
      bsonType: 'object',
      properties: {latitude: 'decimal', longitude: 'decimal'},
      someInvalidProp: {latitude: 0, longitude: 0}
    }
  };
  static plugins = [jsonSchemaPlugin];
}

beforeAll(async () => {
  const db = await mongo.connect(DB_NAME);
  await db.dropDatabase();
});

afterAll(async () => {
  await mongo.disconnect();
});

describe('jsonSchemaPlugin', () => {
  let TestModel: Model;
  it('should properly loadModel', async () => {
    TestModel = await mongo.loadModel(Test);
    expect(TestModel instanceof Model).toBeTruthy();
  });
  it('should properly add `validator.$jsonSchema` to both schema and collection', async () => {
    const {ops, insertedId} = await TestModel.insertOne({name: 'insertOne', list: [1, '2']});
    const {options} = await TestModel.getCollectionInfo();
    expect(options.validator).toHaveProperty('$jsonSchema');
    expect(options.validator.$jsonSchema).toMatchSnapshot();
  });
});
