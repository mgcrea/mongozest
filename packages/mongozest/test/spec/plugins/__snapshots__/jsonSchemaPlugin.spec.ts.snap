// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonSchemaPlugin model with basic props should properly add \`validator.$jsonSchema\` to both schema and collection 1`] = `
Object {
  "additionalProperties": false,
  "bsonType": "object",
  "properties": Object {
    "_id": Object {
      "bsonType": "objectId",
    },
    "avatar": Object {
      "additionalProperties": false,
      "bsonType": "object",
      "properties": Object {
        "fileName": Object {
          "bsonType": "string",
        },
        "size": Object {
          "bsonType": "long",
        },
      },
    },
    "list": Object {
      "bsonType": "array",
      "items": Object {
        "oneOf": Array [
          Object {
            "bsonType": "string",
          },
          Object {
            "bsonType": "int",
          },
        ],
      },
    },
    "name": Object {
      "bsonType": "string",
    },
  },
  "required": Array [
    "name",
  ],
}
`;

exports[`jsonSchemaPlugin model with nested array should properly add \`validator.$jsonSchema\` to both schema and collection 1`] = `
Object {
  "additionalProperties": false,
  "bsonType": "object",
  "properties": Object {
    "_id": Object {
      "bsonType": "objectId",
    },
    "avatar": Object {
      "additionalProperties": false,
      "bsonType": "object",
      "properties": Object {
        "fileName": Object {
          "bsonType": "string",
        },
        "size": Object {
          "bsonType": "long",
        },
      },
    },
    "list": Object {
      "bsonType": "array",
      "items": Object {
        "oneOf": Array [
          Object {
            "bsonType": "string",
          },
          Object {
            "bsonType": "int",
          },
        ],
      },
    },
    "name": Object {
      "bsonType": "string",
    },
  },
  "required": Array [
    "name",
  ],
}
`;

exports[`jsonSchemaPlugin model with nested object should properly add \`validator.$jsonSchema\` to both schema and collection 1`] = `
Object {
  "additionalProperties": false,
  "bsonType": "object",
  "properties": Object {
    "_id": Object {
      "bsonType": "objectId",
    },
    "avatar": Object {
      "additionalProperties": false,
      "bsonType": "object",
      "properties": Object {
        "fileName": Object {
          "bsonType": "string",
        },
        "size": Object {
          "bsonType": "long",
        },
      },
    },
    "list": Object {
      "bsonType": "array",
      "items": Object {
        "oneOf": Array [
          Object {
            "bsonType": "string",
          },
          Object {
            "bsonType": "int",
          },
        ],
      },
    },
    "name": Object {
      "bsonType": "string",
    },
  },
  "required": Array [
    "name",
  ],
}
`;
